### ISSUE TITLE
PersistentActor/ProcessorImpl should use HighSequenceNr to replay events efficiently

### ISSUE THREAD

**Issue by nefilim** [akka/akka#15916]
At the moment it looks like events are replayed from 0 (or snapshot.sequenceNr) to Long.maxValue and only after that is completed a ReadHighestSequenceNr is performed (in ProcessorImpl.onReplaySuccess). 

I propose that ReadHighestSequenceNr is done first and that value is used to:

1. replay events from 0 (or snapshot.sequenceNr) to HighSequenceNr 
2. OR do not attempt to replay events if the HighSequenceNr is 0, probably a new actor 

I believe without these changes it is not viable to use akka-persistence with DynamoDB and it's provisioned throughput model for a "large scale" deployment, especially when the system needs to support a very large number of persistent actors, see my comments & calculations with regard to akka-persistence(-dynamodb) here: https://github.com/sclasen/akka-persistence-dynamodb/issues/3

**Comment by CptnKirk** [comment]
Has this been resolved in the 2.4 milestones?  I just spent a few days tracking down an issue where messages weren't replayed.  The root cause is the journal impl not properly working around this issue for sparse journals.

This feels like a somewhat major bug, and not a nice-to-have optimization.  What's the point of having 

```
def asyncReadHighestSequenceNr(processorId: String, fromSequenceNr: Long): Future[Long] 
```
as part of your journal interface if the framework isn't going to use the value to properly limit the

```
def asyncReplayMessages(processorId: String, fromSequenceNr: Long, toSequenceNr: Long, max: Long)(replayCallback: (PersistentRepr) => Unit): Future[Unit]
```
call?

Now that I know the problem I can fork the journal impl and fix it.  But still, seems like an easy to fix issue that shouldn't take a year to get into a production Akka release.  Sorry for the grump.  Been up for a while tracking this down.

PS:  There should probably be a *ReadLowestSequenceNr call as well.  Without a snapshot, the fromSeq should start at the last non-deleted sequence (unless overridden).  Not 0.  This is also causing havoc in my system.

Imagine a system without snapshots with 1 million records, each with 1 million mutations but coded such that the journal only keeps the last 5 mutations per processorId.  Properly bounding deleteMessages() and replayMessages() is hugely important.  And it's the framework that supplies these bounds.  It's up to the journal impl to faithfully process over them.  And in this case, the framework is being way more expensive than it has to be.  

Sure the journal impl could try and work around issues like these.  But it shouldn't have to, especially not for this long.

**Comment by patriknw** [comment]
@CptnKirk thanks for the update. We should absolutely revisit this part of the journal plugin api, and now is the right time to do so.

**Comment by patriknw** [comment]
The plan is start the replay with asyncReadHighestSequenceNr request if toSequenceNr == Long.MaxValue, and then pass in that result into asyncReplayMessages, and also pass that highestSeqNr in the ReplayMessagesSuccess message to the PersistentActor.

If the asyncReadHighestSequenceNr returns 0 asyncReplayMessages is not called.

I think that is exactly what you suggested.

This is not optimal for PersistentView, but we will deprecate that anyway in favor of the stream based queries.

I don't get the point about lowestSeqNr. We will remove the non-permanent deletion, i.e. there should be no overhead to use 0 as lowestSeqNr.

**Comment by CptnKirk** [comment]
The reason to use the lowSeqNr is the same as providing the high number. It helps the driver properly window requests. The existing Dynamo driver for example will issue batch get and batch delete requests from low-high because that's what was asked of it. 

If high = 100k and low really is 100k - 5 due to deletes, there are tons of noop requests sent to Dynamo if you always start from 0.  Dynamo turns around and eventually responds to them. Costing you consumed resources, throttling kicks in, etc.  The bug was obvious when high = Long.max, so it handled that case. But not so for the lower bounds of the delete and replay calls. 

I've fixed this issue as well in a forked branch. Still, would have been nice not to have had to.  Better values coming from Akka would have helped the journal implementer.

-Jim

-- Sent from my mobile device

**Comment by patriknw** [comment]
@CptnKirk Thanks for clarifying. The lowSeqNr is an optimization that should be handled by the plugin. Akka can't get it from thin air, so it must ask the plugin for it anyway, and that would mean an extra data store query that is unnecessary for some plugins. 

The dynamo plugin should start the `asyncReplayMessages` with getting the lowest sequence number itself.

**Comment by ktoso** [comment]
Moving to `2-4.M3`.

**Comment by CptnKirk** [comment]
I don't understand that logic. The lowSeqNr is the more relevant in most cases since replaying to not-max is rare. What you're saying is that drivers can't trust any values provided by akka and should do their own narrowing in all cases. If that's the case then akka wouldn't need a highSeqNr call from the driver either.  Simply akka should take in values from the user, or detect them from the snapshot system and make the recovery call without any preflight bounding from the journal. 

asyncReplyMessages(atLeastSeqNr, atMostSeqNr)

-Jim

-- Sent from my mobile device

**Comment by patriknw** [comment]
We always need the high sequence number in the PersistentActor as the starting point for persist of new events. Since we already needed it and you said that some journals may benefit from having a defined upper bound in the replay I thought we could move the query of the high sequence number to before the replay, instead of after. That actually simplifies a few things internally in the PersistentActor.

We don't need the low sequence number for anything. The low sequence number is only for optimizing, and it is not a generally applicable optimization (think how you would do it for a SQL database). If a plugin gets a 0L low sequence number it could optimize by retrieving the low sequence number itself. If it is != 0L then the starting point is based on a snapshot.
